File: src-tauri/src/commands/knowledge.rs
// SPDX-License-Identifier: Apache-2.0
use serde::Deserialize;
use tauri::State;

use crate::errors::AppError;
use crate::knowledge::{KnowledgeLibrary, KnowledgeScanSummary};
use crate::AppState;

#[derive(Debug, Clone, Deserialize)]
pub struct RegisterKnowledgeLibraryRequest {
    pub name: String,
    pub root_path: String,
}

#[tauri::command]
pub async fn register_knowledge_library(
    request: RegisterKnowledgeLibraryRequest,
    state: State<'_, AppState>,
) -> Result<KnowledgeLibrary, AppError> {
    let pool = state.db.read().await;
    crate::knowledge::add_library(&pool, request.name, request.root_path).await
}

#[tauri::command]
pub async fn list_knowledge_libraries(
    state: State<'_, AppState>,
) -> Result<Vec<KnowledgeLibrary>, AppError> {
    let pool = state.db.read().await;
    crate::knowledge::list_libraries(&pool).await
}

#[tauri::command]
pub async fn scan_knowledge_library(
    library_id: String,
    state: State<'_, AppState>,
) -> Result<KnowledgeScanSummary, AppError> {
    let pool = state.db.read().await;
    crate::knowledge::scan_library(&pool, &library_id).await
}

#[tauri::command]
pub async fn set_knowledge_library_enabled(
    library_id: String,
    enabled: bool,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    let pool = state.db.read().await;
    crate::knowledge::set_library_enabled(&pool, &library_id, enabled).await
}

#[tauri::command]
pub async fn delete_knowledge_library(
    library_id: String,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    let pool = state.db.read().await;
    crate::knowledge::delete_library(&pool, &library_id).await
}

File: src/stores/knowledge.ts
// SPDX-License-Identifier: Apache-2.0
import { create } from "zustand";
import { invoke } from "../lib/tauri";
import type { KnowledgeLibrary, KnowledgeScanSummary } from "../lib/tauri";

interface KnowledgeState {
  libraries: KnowledgeLibrary[];
  scanSummaries: Record<string, KnowledgeScanSummary>;
  loading: boolean;
  scanning: Record<string, boolean>;
  error: string | null;

  loadLibraries: () => Promise<void>;
  registerLibrary: (name: string, rootPath: string) => Promise<KnowledgeLibrary>;
  scanLibrary: (libraryId: string) => Promise<KnowledgeScanSummary>;
  setLibraryEnabled: (libraryId: string, enabled: boolean) => Promise<void>;
  deleteLibrary: (libraryId: string) => Promise<void>;
}

export const useKnowledgeStore = create<KnowledgeState>((set, get) => ({
  libraries: [],
  scanSummaries: {},
  loading: false,
  scanning: {},
  error: null,

  loadLibraries: async () => {
    set({ loading: true, error: null });
    try {
      const libraries = await invoke<KnowledgeLibrary[]>("list_knowledge_libraries");
      set({ libraries, loading: false });
    } catch (e) {
      set({ loading: false, error: String(e) });
    }
  },

  registerLibrary: async (name, rootPath) => {
    set({ error: null });
    try {
      const library = await invoke<KnowledgeLibrary>("register_knowledge_library", {
        request: { name, root_path: rootPath },
      });
      await get().loadLibraries();
      return library;
    } catch (e) {
      set({ error: String(e) });
      throw e;
    }
  },

  scanLibrary: async (libraryId) => {
    set((s) => ({ scanning: { ...s.scanning, [libraryId]: true }, error: null }));
    try {
      const summary = await invoke<KnowledgeScanSummary>("scan_knowledge_library", {
        libraryId,
      });
      set((s) => ({
        scanSummaries: { ...s.scanSummaries, [libraryId]: summary },
        scanning: { ...s.scanning, [libraryId]: false },
      }));
      await get().loadLibraries();
      return summary;
    } catch (e) {
      set((s) => ({
        scanning: { ...s.scanning, [libraryId]: false },
        error: String(e),
      }));
      throw e;
    }
  },

  setLibraryEnabled: async (libraryId, enabled) => {
    set({ error: null });
    try {
      await invoke("set_knowledge_library_enabled", { libraryId, enabled });
      set((s) => ({
        libraries: s.libraries.map((library) =>
          library.id === libraryId ? { ...library, enabled } : library,
        ),
      }));
    } catch (e) {
      set({ error: String(e) });
      throw e;
    }
  },

  deleteLibrary: async (libraryId) => {
    set({ error: null });
    try {
      await invoke("delete_knowledge_library", { libraryId });
      set((s) => {
        const scanSummaries = { ...s.scanSummaries };
        delete scanSummaries[libraryId];
        return {
          libraries: s.libraries.filter((library) => library.id !== libraryId),
          scanSummaries,
        };
      });
    } catch (e) {
      set({ error: String(e) });
      throw e;
    }
  },
}));
